File: php/src/Domain/Rules/ValueObject/AttributeCode.php
<?php

declare(strict_types=1);

namespace Alxarafe\App\Domain\Rules\ValueObject;

use InvalidArgumentException;

/**
 * Value Object representing an attribute code (e.g. COLD, HAZMAT, FRAGILE).
 */
final readonly class AttributeCode
{
    public function __construct(private string $value)
    {
        if ($value === '') {
            throw new InvalidArgumentException('AttributeCode cannot be empty.');
        }

        if (preg_match('/^[A-Z][A-Z0-9_]{1,19}$/', $value) !== 1) {
            throw new InvalidArgumentException(
                'AttributeCode must be uppercase alphanumeric with underscores (2-20 chars). Got: ' . $value
            );
        }
    }

    public function value(): string
    {
        return $this->value;
    }

    public function equals(self $other): bool
    {
        return $this->value === $other->value;
    }

    public function __toString(): string
    {
        return $this->value;
    }
}

File: java/src/main/java/com/alxarafe/app/domain/rules/valueobject/AttributeCode.java
package com.alxarafe.app.domain.rules.valueobject;

import java.util.Objects;
import java.util.regex.Pattern;

/**
 * Value Object representing an attribute code (e.g. COLD, HAZMAT, FRAGILE).
 */
public record AttributeCode(String value) {

    private static final Pattern FORMAT = Pattern.compile("^[A-Z][A-Z0-9_]{1,19}$");

    public AttributeCode {
        Objects.requireNonNull(value, "AttributeCode cannot be null.");
        if (!FORMAT.matcher(value).matches()) {
            throw new IllegalArgumentException(
                    "AttributeCode must be uppercase alphanumeric with underscores (2-20 chars). Got: " + value);
        }
    }

    @Override
    public String toString() {
        return value;
    }
}
